buildtools/gomodules/gomodules.go

```diff
@@ -0,0 +1,117 @@
+package gomodules
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/fossas/fossa-cli/buildtools"
+	"github.com/fossas/fossa-cli/exec"
+	"github.com/fossas/fossa-cli/pkg"
+)
+
+// Resolver maps modules from import paths to their pkg.import representation.
+type Resolver struct {
+	pathMap map[string]pkg.Import
+}
+
+// module is used to unmarshal `go list -m -json all`.
+type module struct {
+	Path    string
+	Version string
+	Replace replace
+}
+
+type replace struct {
+	Path    string
+	Version string
+}
+
+// Resolve reduces a Go package to its module path and returns its revision
+// contained within pathMap. buildtools.ErrNoRevisionForPackage is returned
+// if the package cannot be found.
+func (r Resolver) Resolve(importpath string) (pkg.Import, error) {
+	splitPath := strings.Split(importpath, "/")
+	for i := range splitPath {
+		revision, ok := r.pathMap[importpath]
+		if ok {
+			return revision, nil
+		}
+		importpath = strings.TrimSuffix(importpath, "/"+splitPath[len(splitPath)-i-1])
+	}
+	return pkg.Import{}, buildtools.ErrNoRevisionForPackage
+}
+
+// New connects goModuleList and a parser to return a golang.Resolver.
+func New(dir string) (Resolver, error) {
+	moduleJSON, err := goModuleList(dir)
+	if err != nil {
+		return Resolver{}, errors.Wrap(err, "Could not run go list")
+	}
+
+	resolver, err := ParseModuleJSON(moduleJSON)
+	if err != nil {
+		return Resolver{}, errors.Wrap(err, "Could not parse json")
+	}
+	return resolver, nil
+}
+
+// ParseModuleJSON returns a golang.Resolver from the output of `go list -m -json all`.
+// Replaced modules are handled in place and added to the pathMap.
+func ParseModuleJSON(moduleJSON string) (Resolver, error) {
```

Comment:
I think that lazily initializing the map makes this approach less straightforward when compared to the executor in pipenv, however, I do think this is a more correct approach. It might be slightly less clear but I'm going to implement this so we establish a pattern to handle the case when we need to call `exec.Run()` within a directory.